File: articles/future-directions-in-systems-modeling/r/future_directions_systems_modeling_workflow.R
# future_directions_systems_modeling_workflow.R
# Base R workflow: streaming observations, rolling updates, and model-drift diagnostics.

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)

if (length(file_arg) > 0) {
  script_path <- normalizePath(sub("^--file=", "", file_arg[1]), mustWork = TRUE)
  article_root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)
} else {
  article_root <- normalizePath(getwd(), mustWork = TRUE)
}

data_dir <- file.path(article_root, "data")
tables_dir <- file.path(article_root, "outputs", "tables")
figures_dir <- file.path(article_root, "outputs", "figures")

dir.create(tables_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(figures_dir, recursive = TRUE, showWarnings = FALSE)

set.seed(42)

n_steps <- 120
time <- 0:(n_steps - 1)

true_state <- numeric(n_steps)
observed_state <- numeric(n_steps)
estimated_state <- numeric(n_steps)
drift_indicator <- numeric(n_steps)
intervention_flag <- numeric(n_steps)
capacity <- rep(18, n_steps)

true_state[1] <- 12
observed_state[1] <- true_state[1] + rnorm(1, 0, 1.0)
estimated_state[1] <- observed_state[1]

for (t in 2:n_steps) {
  shock <- ifelse((t - 1) %in% c(35, 70, 95), 4.5, 0)

  true_state[t] <- 0.93 * true_state[t - 1] +
    0.3 * sin((t - 1) / 10) +
    shock +
    rnorm(1, 0, 0.5)

  observed_state[t] <- true_state[t] + rnorm(1, 0, 1.0)

  prediction <- 0.93 * estimated_state[t - 1] + 0.3 * sin((t - 1) / 10)
  residual <- observed_state[t] - prediction

  if (abs(residual) > 3.0) {
    intervention_flag[t] <- 1
    prediction <- prediction + 0.25 * residual
  }

  estimated_state[t] <- 0.70 * prediction + 0.30 * observed_state[t]

  start_index <- max(1, t - 9)
  drift_indicator[t] <- mean(abs(observed_state[start_index:t] - estimated_state[start_index:t]))
}

df <- data.frame(
  time = time,
  true_state = true_state,
  observed_state = observed_state,
  estimated_state = estimated_state,
  absolute_error_observed = abs(observed_state - true_state),
  absolute_error_estimated = abs(estimated_state - true_state),
  drift_indicator = drift_indicator,
  capacity_margin = capacity - estimated_state,
  intervention_flag = intervention_flag
)

summary_metrics <- data.frame(
  metric = c(
    "MAE_observed",
    "MAE_estimated",
    "Max_drift_indicator",
    "Intervention_count",
    "Minimum_capacity_margin"
  ),
  value = c(
    mean(df$absolute_error_observed),
    mean(df$absolute_error_estimated),
    max(df$drift_indicator),
    sum(df$intervention_flag),
    min(df$capacity_margin)
  )
)

governance_controls <- read.csv(file.path(data_dir, "model_governance_controls.csv"), stringsAsFactors = FALSE)
capability_register <- read.csv(file.path(data_dir, "future_capability_register.csv"), stringsAsFactors = FALSE)
adaptive_triggers <- read.csv(file.path(data_dir, "adaptive_triggers.csv"), stringsAsFactors = FALSE)

validation_checks <- data.frame(
  check = c(
    "time_steps_created",
    "estimated_state_created",
    "observed_errors_nonnegative",
    "estimated_errors_nonnegative",
    "drift_indicator_nonnegative",
    "intervention_flags_binary",
    "governance_controls_created",
    "adaptive_triggers_created"
  ),
  passed = c(
    nrow(df) > 0,
    all(!is.na(df$estimated_state)),
    all(df$absolute_error_observed >= 0),
    all(df$absolute_error_estimated >= 0),
    all(df$drift_indicator >= 0),
    all(df$intervention_flag %in% c(0, 1)),
    nrow(governance_controls) > 0,
    nrow(adaptive_triggers) > 0
  )
)

write.csv(df, file.path(tables_dir, "r_future_systems_modeling_streaming_updates.csv"), row.names = FALSE)
write.csv(summary_metrics, file.path(tables_dir, "r_future_systems_modeling_summary_metrics.csv"), row.names = FALSE)
write.csv(governance_controls, file.path(tables_dir, "r_model_governance_controls.csv"), row.names = FALSE)
write.csv(capability_register, file.path(tables_dir, "r_future_capability_register.csv"), row.names = FALSE)
write.csv(adaptive_triggers, file.path(tables_dir, "r_adaptive_triggers.csv"), row.names = FALSE)
write.csv(validation_checks, file.path(tables_dir, "r_future_systems_modeling_validation_checks.csv"), row.names = FALSE)

png(file.path(figures_dir, "r_streaming_model_updates.png"), width = 1000, height = 700)
plot(
  df$time,
  df$true_state,
  type = "l",
  lwd = 2,
  xlab = "Time",
  ylab = "System State",
  main = "Streaming Observations and Rolling Model Updates"
)
lines(df$time, df$observed_state, lty = 2)
lines(df$time, df$estimated_state, lwd = 2)
points(
  df$time[df$intervention_flag == 1],
  df$estimated_state[df$intervention_flag == 1],
  pch = 19
)
legend(
  "topright",
  legend = c("True State", "Observed State", "Estimated State", "Intervention Trigger"),
  lty = c(1, 2, 1, NA),
  pch = c(NA, NA, NA, 19),
  bty = "n"
)
grid()
dev.off()

print(head(df))
print(summary_metrics)
print(validation_checks)
cat("R future systems modeling workflow complete.\n")

File: articles/future-directions-in-systems-modeling/python/future_directions_systems_modeling_workflow.py
#!/usr/bin/env python3
"""
Future directions in systems modeling.

Dependency-light workflow demonstrating:

1. Hidden system simulation
2. Noisy observations
3. Rolling state estimation
4. Anomaly detection
5. Adaptive intervention flags
6. Drift indicators
7. Governance and trigger registers
8. Validation checks

All data are synthetic.
"""

from __future__ import annotations

import csv
import math
import random
from pathlib import Path


ARTICLE_ROOT = Path(__file__).resolve().parents[1]
DATA = ARTICLE_ROOT / "data"
TABLES = ARTICLE_ROOT / "outputs" / "tables"


def read_csv(path: Path) -> list[dict[str, str]]:
    with path.open("r", newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def write_csv(path: Path, rows: list[dict[str, object]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    if not rows:
        raise ValueError(f"No rows to write: {path}")

    fieldnames: list[str] = []
    for row in rows:
        for key in row:
            if key not in fieldnames:
                fieldnames.append(key)

    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=fieldnames, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(rows)


def mean(values: list[float]) -> float:
    return sum(values) / len(values) if values else 0.0


def simulate_monitoring_loop(n_steps: int = 120) -> list[dict[str, object]]:
    random.seed(42)

    true_state = [0.0] * n_steps
    observed_state = [0.0] * n_steps
    estimated_state = [0.0] * n_steps
    intervention_flag = [0] * n_steps
    drift_indicator = [0.0] * n_steps
    capacity = [18.0] * n_steps

    true_state[0] = 12.0
    observed_state[0] = true_state[0] + random.gauss(0, 1.0)
    estimated_state[0] = observed_state[0]

    for t in range(1, n_steps):
        shock = 4.5 if t in (35, 70, 95) else 0.0

        true_state[t] = (
            0.93 * true_state[t - 1]
            + 0.3 * math.sin(t / 10)
            + shock
            + random.gauss(0, 0.5)
        )

        observed_state[t] = true_state[t] + random.gauss(0, 1.0)

        prediction = 0.93 * estimated_state[t - 1] + 0.3 * math.sin(t / 10)
        residual = observed_state[t] - prediction

        if abs(residual) > 3.0:
            intervention_flag[t] = 1
            prediction = prediction + 0.25 * residual

        estimated_state[t] = 0.70 * prediction + 0.30 * observed_state[t]

        start = max(0, t - 9)
        recent_residuals = [
            abs(observed_state[i] - estimated_state[i])
            for i in range(start, t + 1)
        ]
        drift_indicator[t] = mean(recent_residuals)

    rows: list[dict[str, object]] = []
    for t in range(n_steps):
        rows.append(
            {
                "time": t,
                "true_state": round(true_state[t], 6),
                "observed_state": round(observed_state[t], 6),
                "estimated_state": round(estimated_state[t], 6),
                "absolute_error_observed": round(abs(observed_state[t] - true_state[t]), 6),
                "absolute_error_estimated": round(abs(estimated_state[t] - true_state[t]), 6),
                "drift_indicator": round(drift_indicator[t], 6),
                "capacity_margin": round(capacity[t] - estimated_state[t], 6),
                "intervention_flag": intervention_flag[t],
            }
        )

    return rows


def main() -> None:
    monitoring_rows = simulate_monitoring_loop()
    governance_controls = read_csv(DATA / "model_governance_controls.csv")
    capability_register = read_csv(DATA / "future_capability_register.csv")
    adaptive_triggers = read_csv(DATA / "adaptive_triggers.csv")

    observed_errors = [float(row["absolute_error_observed"]) for row in monitoring_rows]
    estimated_errors = [float(row["absolute_error_estimated"]) for row in monitoring_rows]
    drift_values = [float(row["drift_indicator"]) for row in monitoring_rows]
    interventions = [int(row["intervention_flag"]) for row in monitoring_rows]

    summary_rows = [
        {"metric": "MAE_observed", "value": round(mean(observed_errors), 6)},
        {"metric": "MAE_estimated", "value": round(mean(estimated_errors), 6)},
        {"metric": "Max_drift_indicator", "value": round(max(drift_values), 6)},
        {"metric": "Intervention_count", "value": sum(interventions)},
        {"metric": "Minimum_capacity_margin", "value": round(min(float(row["capacity_margin"]) for row in monitoring_rows), 6)},
    ]

    validation_rows = [
        {"check": "time_steps_created", "passed": len(monitoring_rows) > 0, "value": len(monitoring_rows)},
        {"check": "estimated_state_created", "passed": all(row["estimated_state"] is not None for row in monitoring_rows), "value": "all_estimates_checked"},
        {"check": "observed_errors_nonnegative", "passed": all(float(row["absolute_error_observed"]) >= 0 for row in monitoring_rows), "value": "all_observed_errors_checked"},
        {"check": "estimated_errors_nonnegative", "passed": all(float(row["absolute_error_estimated"]) >= 0 for row in monitoring_rows), "value": "all_estimated_errors_checked"},
        {"check": "drift_indicator_nonnegative", "passed": all(float(row["drift_indicator"]) >= 0 for row in monitoring_rows), "value": "all_drift_indicators_checked"},
        {"check": "intervention_flags_binary", "passed": all(row["intervention_flag"] in (0, 1) for row in monitoring_rows), "value": "all_intervention_flags_checked"},
        {"check": "governance_controls_created", "passed": len(governance_controls) > 0, "value": len(governance_controls)},
        {"check": "adaptive_triggers_created", "passed": len(adaptive_triggers) > 0, "value": len(adaptive_triggers)},
    ]

    write_csv(TABLES / "python_future_systems_modeling_hybrid_monitoring.csv", monitoring_rows)
    write_csv(TABLES / "python_future_systems_modeling_hybrid_summary.csv", summary_rows)
    write_csv(TABLES / "python_model_governance_controls.csv", governance_controls)
    write_csv(TABLES / "python_future_capability_register.csv", capability_register)
    write_csv(TABLES / "python_adaptive_triggers.csv", adaptive_triggers)
    write_csv(TABLES / "python_future_systems_modeling_validation_checks.csv", validation_rows)

    print("Future systems modeling hybrid monitoring workflow complete.")
    print(TABLES / "python_future_systems_modeling_hybrid_monitoring.csv")


if __name__ == "__main__":
    main()
